apps/ejabberd/src/mod_privacy_riak.erl

```diff
@@ -0,0 +1,166 @@
+%%==============================================================================
+
+%
+% NOTES:
+%   user default privacy is stored in bucket {<<"privacy_defaults">>, LServer}, key LUser
+%   user privacy lists names are stored in set {<<"privacy_lists_names">>, <<"LServer">>}, key LUser
+%   user privace lists content are stored in bucket {<<"privacy_lists">>, LServer}, key <<LUser,$/,ListName>>
+
+-module(mod_privacy_riak).
+-author('guillaume@bour.cc').
+-behaviour(mod_privacy).
+
+-export([init/2,
+         get_default_list/2,
+         get_list_names/2,
+         get_privacy_list/3,
+         forget_default_list/2,
+         set_default_list/3,
+         remove_privacy_list/3,
+         replace_privacy_list/4,
+         remove_user/2]).
+
+-include("ejabberd.hrl").
+-include("jlib.hrl").
+-include("mod_privacy.hrl").
+
+-include_lib("riakc/include/riakc.hrl").
+
+-define(BKT_DEFAULT_LIST(Server), {<<"privacy_defaults">>, Server}).
+-define(BKT_LISTS_NAMES(Server), {<<"privacy_lists_names">>, Server}).
+-define(BKT_LISTS(Server), {<<"privacy_lists">>, Server}).
+
+init(_Host, _Opts) ->
+    ok.
+
+get_default_list(LUser, LServer) ->
+    case get_default_list_name(LUser, LServer) of
+        none ->
+            {error, not_found};
+        Default ->
+            case get_privacy_list(LUser, LServer, Default) of
+                {ok, List} ->
+                    {ok, {Default, List}};
+                {error, Reason} ->
+                    {error, Reason}
+            end
+    end.
+
+get_list_names(LUser, LServer) ->
+    Default = get_default_list_name(LUser, LServer),
+    Names = get_list_names_only(LUser, LServer),
+    {ok, {Default, Names}}.
+
+get_default_list_name(LUser, LServer) ->
+    case mongoose_riak:get(?BKT_DEFAULT_LIST(LServer), LUser) of
+        {ok, Obj} ->
+            riakc_obj:get_value(Obj);
+
+        _ -> none
+    end.
+
+get_list_names_only(LUser, LServer) ->
+    case mongoose_riak:fetch_type(?BKT_LISTS_NAMES(LServer), LUser) of
+        {ok, Set} ->
+            riakc_set:value(Set);
+
+        {error, {notfound, set}} ->
+            [];
+
+        Err -> 
+            lager:error("~p", [Err]),
+            []
+    end.
+
+get_privacy_list(LUser, LServer, Name) ->
+    case mongoose_riak:get(?BKT_LISTS(LServer), <<LUser/binary, $/, Name/binary>>) of
+        {ok, Obj} ->
+            Val = binary_to_term(riakc_obj:get_value(Obj)),
+            {ok, Val};
+
+        {error, notfound} ->
+            {error, not_found};
+
+        Err->
+            lager:error("~p", [Err]),
+            Err
+    end.
+
+forget_default_list(LUser, LServer) ->
+    mongoose_riak:delete(?BKT_DEFAULT_LIST(LServer), LUser).
+
+set_default_list(LUser, LServer, Name) ->
+    case mongoose_riak:get(?BKT_LISTS(LServer), <<LUser/binary, $/, Name/binary>>) of
+        {ok, _} ->
+            case mongoose_riak:get(?BKT_DEFAULT_LIST(LServer), LUser) of
+                % update entry
+                {ok, Obj} ->
```

Comment:
There is no need to distinguish between updating and creating new entry as in the end last write wins IMO.